File: frontend/flutter/lib/models/nutrition_profile.dart
/// 营养推荐档案模型类
/// 与账号本身分开，一个账号可建立多个档案（自己、家人等）
class NutritionProfile {
  final String id;
  final String user_id;
  final String name;
  final String? description;
  final double? height;
  final double? weight;
  final int? age;
  final String? gender;
  final String? activityLevel;
  final List<String>? healthConditions;
  final String? goals;
  final List<String>? allergies;
  final List<String>? avoidedIngredients;
  final String? cuisinePreference;
  final String? spicyPreference;
  final DietaryPreferences? dietaryPreferences;
  final bool hasCompletedHealthInfo;
  final DateTime createdAt;
  final DateTime updatedAt;

  NutritionProfile({
    required this.id,
    required this.user_id,
    required this.name,
    this.description,
    this.height,
    this.weight,
    this.age,
    this.gender,
    this.activityLevel,
    this.healthConditions,
    this.goals,
    this.allergies,
    this.avoidedIngredients,
    this.cuisinePreference,
    this.spicyPreference,
    this.dietaryPreferences,
    this.hasCompletedHealthInfo = false,
    required this.createdAt,
    required this.updatedAt,
  });

  // 计算BMI指数
  double? get bmi {
    if (height != null && weight != null && height! > 0) {
      return weight! / ((height! / 100) * (height! / 100));
    }
    return null;
  }

  // 从JSON创建NutritionProfile对象
  factory NutritionProfile.fromJson(Map<String, dynamic> json) {
    return NutritionProfile(
      id: json['_id'] ?? json['id'] ?? '',
      user_id: json['user_id'] ?? json['ownerId'] ?? '',
      name: json['name'] ?? '',
      description: json['description'],
      height: json['height']?.toDouble(),
      weight: json['weight']?.toDouble(),
      age: json['age'],
      gender: json['gender'],
      activityLevel: json['activityLevel'] ?? json['activity_level'],
      healthConditions: json['healthConditions'] != null 
          ? List<String>.from(json['healthConditions']) 
          : null,
      goals: json['goals'],
      allergies: json['allergies'] != null 
          ? List<String>.from(json['allergies']) 
          : null,
      avoidedIngredients: json['avoidedIngredients'] != null 
          ? List<String>.from(json['avoidedIngredients']) 
          : null,
      cuisinePreference: json['cuisinePreference'] ?? json['cuisine_preference'],
      spicyPreference: json['spicyPreference'] ?? json['spicy_preference'],
      dietaryPreferences: json['dietaryPreferences'] != null 
          ? DietaryPreferences.fromJson(json['dietaryPreferences']) 
          : (json['dietary_preferences'] != null
              ? DietaryPreferences.fromJson(json['dietary_preferences'])
              : null),
      hasCompletedHealthInfo: json['hasCompletedHealthInfo'] ?? false,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : (json['created_at'] != null 
              ? DateTime.parse(json['created_at']) 
              : DateTime.now()),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : (json['updated_at'] != null 
              ? DateTime.parse(json['updated_at']) 
              : DateTime.now()),
    );
  }

  // 将NutritionProfile对象转换为JSON
  Map<String, dynamic> toJson() {
    final map = {
      '_id': id,
      'name': name,
      'user_id': user_id,
      'description': description,
      'height': height,
      'weight': weight,
      'age': age,
      'gender': gender,
      'activityLevel': activityLevel,
      'healthConditions': healthConditions,
      'goals': goals,
      'hasCompletedHealthInfo': hasCompletedHealthInfo,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
    
    // 如果有饮食偏好，则添加饮食偏好对象
    if (dietaryPreferences != null) {
      map['dietaryPreferences'] = dietaryPreferences!.toJson();
    } else {
      // 如果没有饮食偏好对象但有单独的偏好字段，则创建饮食偏好对象
      if (cuisinePreference != null || spicyPreference != null || allergies != null || avoidedIngredients != null) {
        map['dietaryPreferences'] = {
          'cuisinePreference': cuisinePreference,
          'spicyPreference': spicyPreference,
          'allergies': allergies,
          'avoidedIngredients': avoidedIngredients,
        };
      }
    }
    
    return map;
  }

  // 返回营养档案的调试信息
  @override
  String toString() {
    return 'NutritionProfile{id: $id, name: $name, user_id: $user_id, '
           'height: $height, weight: $weight, age: $age}';
  }

  // 创建带有更新属性的NutritionProfile副本
  NutritionProfile copyWith({
    String? id,
    String? name,
    String? user_id,
    String? description,
    double? height,
    double? weight,
    int? age,
    String? gender,
    String? activityLevel,
    List<String>? healthConditions,
    String? goals,
    List<String>? allergies,
    List<String>? avoidedIngredients,
    String? cuisinePreference,
    String? spicyPreference,
    DietaryPreferences? dietaryPreferences,
    bool? hasCompletedHealthInfo,
  }) {
    return NutritionProfile(
      id: id ?? this.id,
      user_id: user_id ?? this.user_id,
      name: name ?? this.name,
      description: description ?? this.description,
      height: height ?? this.height,
      weight: weight ?? this.weight,
      age: age ?? this.age,
      gender: gender ?? this.gender,
      activityLevel: activityLevel ?? this.activityLevel,
      healthConditions: healthConditions ?? this.healthConditions,
      goals: goals ?? this.goals,
      allergies: allergies ?? this.allergies,
      avoidedIngredients: avoidedIngredients ?? this.avoidedIngredients,
      cuisinePreference: cuisinePreference ?? this.cuisinePreference,
      spicyPreference: spicyPreference ?? this.spicyPreference,
      dietaryPreferences: dietaryPreferences ?? this.dietaryPreferences,
      hasCompletedHealthInfo: hasCompletedHealthInfo ?? this.hasCompletedHealthInfo,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
}

class DietaryPreferences {
  final String? cuisinePreference;
  final String? spicyPreference;
  final List<String>? allergies;
  final List<String>? avoidedIngredients;

  DietaryPreferences({
    this.cuisinePreference,
    this.spicyPreference,
    this.allergies,
    this.avoidedIngredients,
  });

  factory DietaryPreferences.fromJson(Map<String, dynamic> json) {
    return DietaryPreferences(
      cuisinePreference: json['cuisinePreference'],
      spicyPreference: json['spicyPreference'],
      allergies: json['allergies'] != null
          ? List<String>.from(json['allergies'])
          : null,
      avoidedIngredients: json['avoidedIngredients'] != null
          ? List<String>.from(json['avoidedIngredients'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'cuisinePreference': cuisinePreference,
      'spicyPreference': spicyPreference,
      'allergies': allergies,
      'avoidedIngredients': avoidedIngredients,
    };
  }
}

File: frontend/src/components/NutritionProfile.js
import React, { useState, useEffect } from 'react';
import { nutritionProfileApi } from '../services/api';
import { useAuth } from '../contexts/AuthContext';
import {
    Box,
    Card,
    CardContent,
    Typography,
    Button,
    TextField,
    Grid,
    Select,
    MenuItem,
    FormControl,
    InputLabel,
    Chip,
    Alert,
    CircularProgress
} from '@mui/material';

const NutritionProfile = () => {
    const { user } = useAuth();
    const [profiles, setProfiles] = useState([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);
    const [editingProfile, setEditingProfile] = useState(null);
    const [formData, setFormData] = useState({
        name: '',
        gender: 'other',
        age: '',
        height: '',
        weight: '',
        activity_level: 'moderate',
        health_conditions: [],
        dietary_preferences: {
            cuisine_preference: 'other',
            allergies: [],
            avoided_ingredients: [],
            spicy_preference: 'medium',
            diet_type: 'omnivore'
        },
        goals: 'health_improvement',
        nutrition_targets: {
            calories: '',
            protein_percentage: '',
            carbs_percentage: '',
            fat_percentage: ''
        }
    });

    useEffect(() => {
        loadProfiles();
    }, [user]);

    const loadProfiles = async () => {
        try {
            setLoading(true);
            const response = await nutritionProfileApi.getUserProfiles(user.id);
            setProfiles(response.profiles);
            setError(null);
        } catch (err) {
            setError('加载营养档案失败');
            console.error(err);
        } finally {
            setLoading(false);
        }
    };

    const handleSubmit = async (e) => {
        e.preventDefault();
        try {
            if (editingProfile) {
                await nutritionProfileApi.updateProfile(editingProfile._id, formData);
            } else {
                await nutritionProfileApi.createProfile(formData);
            }
            await loadProfiles();
            setEditingProfile(null);
            setFormData({
                name: '',
                gender: 'other',
                age: '',
                height: '',
                weight: '',
                activity_level: 'moderate',
                health_conditions: [],
                dietary_preferences: {
                    cuisine_preference: 'other',
                    allergies: [],
                    avoided_ingredients: [],
                    spicy_preference: 'medium',
                    diet_type: 'omnivore'
                },
                goals: 'health_improvement',
                nutrition_targets: {
                    calories: '',
                    protein_percentage: '',
                    carbs_percentage: '',
                    fat_percentage: ''
                }
            });
        } catch (err) {
            setError('保存营养档案失败');
            console.error(err);
        }
    };

    const handleEdit = (profile) => {
        setEditingProfile(profile);
        setFormData(profile);
    };

    const handleDelete = async (profileId) => {
        if (window.confirm('确定要删除这个营养档案吗？')) {
            try {
                await nutritionProfileApi.deleteProfile(profileId);
                await loadProfiles();
            } catch (err) {
                setError('删除营养档案失败');
                console.error(err);
            }
        }
    };

    if (loading) {
        return (
            <Box display="flex" justifyContent="center" alignItems="center" minHeight="200px">
                <CircularProgress />
            </Box>
        );
    }

    return (
        <Box>
            {error && (
                <Alert severity="error" sx={{ mb: 2 }}>
                    {error}
                </Alert>
            )}

            <Card sx={{ mb: 4 }}>
                <CardContent>
                    <Typography variant="h6" gutterBottom>
                        {editingProfile ? '编辑营养档案' : '创建新营养档案'}
                    </Typography>
                    <form onSubmit={handleSubmit}>
                        <Grid container spacing={2}>
                            <Grid item xs={12} md={6}>
                                <TextField
                                    fullWidth
                                    label="档案名称"
                                    value={formData.name}
                                    onChange={(e) => setFormData({ ...formData, name: e.target.value })}
                                    required
                                />
                            </Grid>
                            <Grid item xs={12} md={6}>
                                <FormControl fullWidth>
                                    <InputLabel>性别</InputLabel>
                                    <Select
                                        value={formData.gender}
                                        onChange={(e) => setFormData({ ...formData, gender: e.target.value })}
                                        label="性别"
                                    >
                                        <MenuItem value="male">男</MenuItem>
                                        <MenuItem value="female">女</MenuItem>
                                        <MenuItem value="other">其他</MenuItem>
                                    </Select>
                                </FormControl>
                            </Grid>
                            <Grid item xs={12} md={4}>
                                <TextField
                                    fullWidth
                                    label="年龄"
                                    type="number"
                                    value={formData.age}
                                    onChange={(e) => setFormData({ ...formData, age: e.target.value })}
                                />
                            </Grid>
                            <Grid item xs={12} md={4}>
                                <TextField
                                    fullWidth
                                    label="身高 (cm)"
                                    type="number"
                                    value={formData.height}
                                    onChange={(e) => setFormData({ ...formData, height: e.target.value })}
                                />
                            </Grid>
                            <Grid item xs={12} md={4}>
                                <TextField
                                    fullWidth
                                    label="体重 (kg)"
                                    type="number"
                                    value={formData.weight}
                                    onChange={(e) => setFormData({ ...formData, weight: e.target.value })}
                                />
                            </Grid>
                            <Grid item xs={12}>
                                <FormControl fullWidth>
                                    <InputLabel>活动水平</InputLabel>
                                    <Select
                                        value={formData.activity_level}
                                        onChange={(e) => setFormData({ ...formData, activity_level: e.target.value })}
                                        label="活动水平"
                                    >
                                        <MenuItem value="sedentary">久坐不动</MenuItem>
                                        <MenuItem value="light">轻度活动</MenuItem>
                                        <MenuItem value="moderate">中度活动</MenuItem>
                                        <MenuItem value="active">活跃</MenuItem>
                                        <MenuItem value="very_active">非常活跃</MenuItem>
                                    </Select>
                                </FormControl>
                            </Grid>
                            <Grid item xs={12}>
                                <FormControl fullWidth>
                                    <InputLabel>饮食目标</InputLabel>
                                    <Select
                                        value={formData.goals}
                                        onChange={(e) => setFormData({ ...formData, goals: e.target.value })}
                                        label="饮食目标"
                                    >
                                        <MenuItem value="weight_loss">减重</MenuItem>
                                        <MenuItem value="weight_gain">增重</MenuItem>
                                        <MenuItem value="maintenance">维持体重</MenuItem>
                                        <MenuItem value="muscle_gain">增肌</MenuItem>
                                        <MenuItem value="health_improvement">改善健康</MenuItem>
                                    </Select>
                                </FormControl>
                            </Grid>
                            <Grid item xs={12}>
                                <Box display="flex" gap={2}>
                                    <Button
                                        type="submit"
                                        variant="contained"
                                        color="primary"
                                        fullWidth
                                    >
                                        {editingProfile ? '保存更改' : '创建档案'}
                                    </Button>
                                    {editingProfile && (
                                        <Button
                                            variant="outlined"
                                            color="secondary"
                                            onClick={() => {
                                                setEditingProfile(null);
                                                setFormData({
                                                    name: '',
                                                    gender: 'other',
                                                    age: '',
                                                    height: '',
                                                    weight: '',
                                                    activity_level: 'moderate',
                                                    health_conditions: [],
                                                    dietary_preferences: {
                                                        cuisine_preference: 'other',
                                                        allergies: [],
                                                        avoided_ingredients: [],
                                                        spicy_preference: 'medium',
                                                        diet_type: 'omnivore'
                                                    },
                                                    goals: 'health_improvement',
                                                    nutrition_targets: {
                                                        calories: '',
                                                        protein_percentage: '',
                                                        carbs_percentage: '',
                                                        fat_percentage: ''
                                                    }
                                                });
                                            }}
                                        >
                                            取消编辑
                                        </Button>
                                    )}
                                </Box>
                            </Grid>
                        </Grid>
                    </form>
                </CardContent>
            </Card>

            <Typography variant="h6" gutterBottom>
                我的营养档案
            </Typography>
            <Grid container spacing={2}>
                {profiles.map((profile) => (
                    <Grid item xs={12} md={6} key={profile._id}>
                        <Card>
                            <CardContent>
                                <Typography variant="h6">{profile.name}</Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    性别: {profile.gender === 'male' ? '男' : profile.gender === 'female' ? '女' : '其他'}
                                </Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    年龄: {profile.age}
                                </Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    身高: {profile.height}cm
                                </Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    体重: {profile.weight}kg
                                </Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    活动水平: {
                                        {
                                            sedentary: '久坐不动',
                                            light: '轻度活动',
                                            moderate: '中度活动',
                                            active: '活跃',
                                            very_active: '非常活跃'
                                        }[profile.activity_level]
                                    }
                                </Typography>
                                <Typography color="textSecondary" gutterBottom>
                                    目标: {
                                        {
                                            weight_loss: '减重',
                                            weight_gain: '增重',
                                            maintenance: '维持体重',
                                            muscle_gain: '增肌',
                                            health_improvement: '改善健康'
                                        }[profile.goals]
                                    }
                                </Typography>
                                <Box mt={2} display="flex" gap={1}>
                                    <Button
                                        variant="outlined"
                                        size="small"
                                        onClick={() => handleEdit(profile)}
                                    >
                                        编辑
                                    </Button>
                                    <Button
                                        variant="outlined"
                                        color="error"
                                        size="small"
                                        onClick={() => handleDelete(profile._id)}
                                    >
                                        删除
                                    </Button>
                                </Box>
                            </CardContent>
                        </Card>
                    </Grid>
                ))}
            </Grid>
        </Box>
    );
};

export default NutritionProfile; 